clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1108,13 +1165,39 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
     return isValid();
   }
 
+  bool handlePointerType(FieldDecl *FD, QualType FieldTy) final {
+    // TODO: Replace with a better diagnostic.
+    if (UnionCount) {
```

Comment:
I considered making this a separate handler that only overloaded enter/leave Union and the handleSyclAccessorType/handleSyclPointerType.

We might also consider still doing this, and putting it in after the normal FieldChecker.  Note the bad error messages.